### DISCUSSION TITLE
More pythonic wrapper for libpq ?

### DISCUSSION THREAD

**Discussion by philpep** [psycopg/psycopg#204] [Q&A]
Hi thanks for this new psycopg!

I'm writing some code to implement client side encryption of password using PQencryptPasswordConn, the resulting code looks like:

```python
import psycopg.pq
import psycopg.pq.abc
import psycopg.sql as sql


def encrypt_password(conn: psycopg.pq.abc.PGconn, role: str, password: str) -> str:
    return conn.encrypt_password(password.encode(), role.encode()).decode()

with psycopg.connect() as c:
    c.execute(
        sql.SQL("ALTER ROLE foo WITH PASSWORD {}").format(
            sql.Literal(encrypt_password(c.pgconn, "foo", "secret"))
        )
    )

```

I wonder If we could get rid of encode(), decode() and have an intermediate wrapper on top of libpq which could be more "pythonic".
Also it would be nice to type with `psycopg.pq.PGconn` instead of `psycopg.pq.abc.PGConn`.

What do you think?

**Comment by dvarrazzo** [comment]
Hello,

> Hi thanks for this new psycopg!

:upside_down_face: 

> I wonder If we could get rid of encode(), decode() and have an intermediate wrapper on top of libpq which could be more "pythonic".

the underlying PGconn is a thin wrapper around the libpq c calls, so it's bytes-based.  Its aim is to be complete, more than Pythonic (it is already pythonised in error management etc.)

Psycopg builds on top of it, and typically has a concept of current encoding (which is the connection client encoding) and is typically string-based. For instance:

```
In [4]: conn = psycopg.connect()

In [8]: conn.pgconn.user
Out[8]: b'piro'

In [10]: conn.info.user
Out[10]: 'piro'
```

I'm open to "pythonise more", and I think that psycopg is the right layer to do it, not psycopg.pq. However, I wouldn't like to bloat the connection interface too much. So, where could we put this helper function? on `ConnectionInfo` (i.e. `conn.info.enctypt_password()`) doesn't feel right either.

Note: your helper function assumes utf-8 encoding. A better implementation is probably:

```python
def encrypt_password(conn: psycopg.Connection, role: str, password: str) -> str:
    enc = conn.info.encoding
    return conn.pgconn.encrypt_password(password.encode(enc), role.encode(enc)).decode(enc)
```

----

> Also it would be nice to type with psycopg.pq.PGconn instead of psycopg.pq.abc.PGConn.

`psycopg.pq.PGconn` is an object re-exposed, which might be imported from `psycopg.pq.pq_ctypes.PGconn`, `psycopg_c.pq.PGconn`, `psycopg_binary.pq.PGconn`. They all share the same interface: `psycopg.pq.abc.PGconn`. I am not sure that we can make mypy happy by having `psycopg.pq.PGconn` to be both a concrete object and an interface for the other implementations.

Also note that the `__module__` attribute for `psycopg.pq.pq_ctypes.PGconn` is customized to `psycopg.pq.PGconn`: this is probably not helpful: I think the customization should be removed.

Open to listen and to improve on both the points :)

**Reply by dlax** [reply]
> I'm open to "pythonise more", and I think that psycopg is the right layer to do it, not psycopg.pq. However, I wouldn't like to bloat the connection interface too much. So, where could we put this helper function? on `ConnectionInfo` (i.e. `conn.info.enctypt_password()`) doesn't feel right either.

Also, it seems that there should be an `async` version of that `encrypt_password()` function so we should take care about making them both available from different submodules. Unless I missed something there's no such modules already, so we'd need a name for these... Do we want something like `psycopg.extensions` and `psycopg.async_extensions`? In fact, a method on `[Async]Connection` does not seem inappropriate to me...

**Reply by dvarrazzo** [reply]
Should there? It doesn't that there is an async version of `PQencryptPasswordConn()` to use...

**Reply by dlax** [reply]
I mentioned that because it seems that `PQencryptPasswordConn()` does I/O; but since it's outside of our control, maybe it does not matter?

**Comment by dvarrazzo** [comment]
> Also note that the `__module__` attribute for `psycopg.pq.pq_ctypes.PGconn` is customized to `psycopg.pq.PGconn`: this is probably not helpful: I think the customization should be removed.

Changing this. Now it should be more obvious that, when you use the interface, `psycopg.pq.abc.PGconn` you may refer to `psycopg.pq.pq_ctypes.PGconn`, `psycopg_c.pq.PGconn` or to any other object implementing the same interface.

It still needs documentation though...

**Comment by DylanYoung** [comment]
I think the only IO it does is when `algorithm` is NULL, which causes it to query the server with `SHOW password_encryption`, so an async version could run this query separately and pass the actual algorithm to  `PQencryptPasswordConn`.  

As for where to put it.... If this is the only utility function of this nature, I'd probably just put it on the Connection class, but if there are more functions that would be good candidates, maybe a separate bundling class assigned to `Connection.tool` or some other (bikesheddable) name.   I guess one other candidate might be `change_password`, though it feels like that one should live on the Connection (if implemented).  Another alternative might be a separate `Utility` or `Admin` class (again bikesheddable), that takes a connection in  `__init__` and exposes convenience methods for administrative commands (in the future it might hold things like a `show` method, common monitoring queries, etc.) or this could be a connection subclass.

I guess the simplest option would just be a top-level module that holds bare functions `psycopg.tool` or `psycopg.admin` or something.  

This seems relatively straight-forward and I'd be happy to take a crack at an MR if we can decide on a good spot to put it.

**Reply by dvarrazzo** [reply]
I would also think that an external object taking a connection as argument would be a sensible choice in order to avoid adding arms and legs to the already rich connection interface.

Objects that take a connection to use for querying are tricky. We have examples with `TypeInfo` and related subclasses and they are pretty hairy as the try to guarantee to return a connection in the state they were given, so they have to manage transactions, cursor types, non-idle state when given, errors when returned, to work on sync/async connections.

So, in this case, where the function blocks if a certain argument is not given, and it is documented how to obtain that argument, my preference would largely to just document the behaviour and invite the user to query the database on their own and pass an argument to the function, which is the same that the libpq [suggests in their docs](https://www.postgresql.org/docs/current/libpq-misc.html#LIBPQ-PQENCRYPTPASSWORDCONN):

> If you wish to use the default algorithm for the server but want to avoid blocking, query password_encryption yourself before calling [PQencryptPasswordConn](https://www.postgresql.org/docs/current/libpq-misc.html#LIBPQ-PQENCRYPTPASSWORDCONN), and pass that value as the algorithm.

I don't know how pressing is the request to support the password-related functions with a higher level interface. AFAICS the only thing that really requires the libpq is `PQencryptPasswordConn()` with algorithm specified, whereas `PQchangePassword()` just seems to run a query (looking at the docs, not at the code) and can be easily done in Python. It has all the problems above w.r.t. using a connection for querying, and on top of that the transaction management would be trickier, so I'd really prefer to leave running this query under the responsibility of the user.

**Reply by DylanYoung** [reply]
Yeah.  That's what I was thinking as well.
